lib/modules.nix

```diff
@@ -475,26 +474,54 @@ rec {
           [{ inherit (module) file; inherit value; }]
         ) configs;
 
+      # Convert an option tree decl to a submodule option decl
+      optionTreeToOption = decl:
+        if isOption decl.options
+        then decl
+        else decl // {
+            options = mkOption {
+              type = types.submoduleWith {
+                modules = [ { options = decl.options; } ];
+              };
+            };
+          };
+
       resultsByName = mapAttrs (name: decls:
         # We're descending into attribute ‘name’.
         let
           loc = prefix ++ [name];
           defns = defnsByName.${name} or [];
           defns' = defnsByName'.${name} or [];
-          nrOptions = count (m: isOption m.options) decls;
+          optionDecls = filter (m: isOption m.options) decls;
         in
-          if nrOptions == length decls then
+          if length optionDecls == length decls then
             let opt = fixupOptionType loc (mergeOptionDecls loc decls);
             in {
               matchedOptions = evalOptionValue loc opt defns';
               unmatchedDefns = [];
             }
-          else if nrOptions != 0 then
-            let
-              firstOption = findFirst (m: isOption m.options) "" decls;
-              firstNonOption = findFirst (m: !isOption m.options) "" decls;
-            in
-              throw "The option `${showOption loc}' in `${firstOption._file}' is a prefix of options in `${firstNonOption._file}'."
+          else if optionDecls != [] then
+              if (lib.head optionDecls).options.type.name == "submodule"
```

Comment:
This leads to a different error depending on module order. E.g.
```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith { modules = []; };
      };
    }
    {
      _file = "b.nix";
      options.foo = lib.mkOption {
        type = lib.types.int;
      };
    }
    {
      _file = "c.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: The option `foo' in `b.nix' is a prefix of options in `c.nix'.
(use '--show-trace' to show detailed location information)
```

While

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "b.nix";
      options.foo = lib.mkOption {
        type = lib.types.int;
      };
    }
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith { modules = []; };
      };
    }
    {
      _file = "c.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: The option `foo' in `b.nix' is already declared in `c.nix' and `a.nix'.
(use '--show-trace' to show detailed location information)
```

Not yet sure what the best error to throw is in such a case

Reply:
Yeah, I've inherited this flaw from the existing code.
Might have a look later, but not sure if this needs to be scoped in.

Reply:
Previously the order of modules only had an influence on the specific files mentioned in the error message, but the error message stayed the same. This PR would make the error message variety itself change.

Reply:
It now checks all and I've improved the error message.